avocado/utils/vmimage.py

```diff
@@ -0,0 +1,375 @@
+import string
+import tempfile
+import urllib2
+import uuid
+
+try:
+    import lzma
+    LZMA_CAPABLE = True
+except ImportError:
+    LZMA_CAPABLE = False
+
+from HTMLParser import HTMLParser
+
+from . import asset
+from . import path as utils_path
+from . import process
+
+
+class ImageProviderError(Exception):
+    """
+    Generic error class for ImageProvider
+    """
+
+
+class VMImageHtmlParser(HTMLParser):
+    def __init__(self, pattern):
+        HTMLParser.__init__(self)
+        self.versions = []
+        self.pattern = pattern
+
+    def handle_starttag(self, tag, attrs):
+        if tag != 'a':
+            return
+        for attr in attrs:
+            if attr[0] == 'href' and re.match(self.pattern, attr[1]):
+                self.versions.append(attr[1].strip('/'))
+
+
+class ImageBase(object):
+    def __init__(self, **kwargs):
+        self.path = None
+        self.url = None
+
+        self.url_versions = None
+        self.url_images = None
+        self.image = None
+
+        self.default_version = None
+        self.default_build = None
+        self.default_arch = None
+
+        self.args = {}
+        self.initargs = kwargs
+
+    def get(self):
+        self.url = self.get_url()
+
+        asset_path = asset.Asset(name=self.url,
+                                 asset_hash=self.args['checksum'],
+                                 algorithm=self.args['algorithm'],
+                                 locations=None,
+                                 cache_dirs=[self.args['cache_dir']],
+                                 expire=None).fetch()
+
+        if os.path.splitext(asset_path)[1] == '.xz':
+            if LZMA_CAPABLE:
+                asset_path = self._extract(asset_path)
+            else:
+                raise ImageProviderError('Cannot uncompress image')
+
+        self.path = self._create_backed_image(asset_path)
+        return self.path
+
+    def get_url(self):
+        # Did user pass all the args required to create the URL?
+        image_url = self.url_images + self.image
+        required_args = []
+        for key in string.Formatter().parse(image_url):
+            if key[1] is not None:
+                required_args.append(key[1])
+        all_args = True
+        for required_arg in required_args:
+            if (required_arg not in self.initargs or
+                    self.initargs[required_arg] is None):
+                all_args = False
+        # In that case, URL is ready
+        if all_args:
+            return image_url.format(**self.initargs)
+
+        # Otherwise, we have to probe it externally
+        # Getting the requested version, if available
+        version = self.get_version()
+        # Getting the image URL
+        return self.get_image(version)
+
+    def get_version(self):
+        pattern = '^%s/$' % self.args['version']
+
+        parser = VMImageHtmlParser(pattern)
+        try:
+            parser.feed(urllib2.urlopen(self.url_versions).read())
+        except:
```

Comment:
naked `except` should be used only exceptionally. Here I don't see reason for it as `urllib` reports a useful message. If you insist on wrapping the exception you should only catch the `urllib2.URLError` (or similar)

Reply:
Ack. Catching the specific exception.